### PR TITLE
Revocable memory POC

### PR DESCRIPTION
**Pull request by pnowojski**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java [hunk 1]
```diff
@@ -72,16 +83,47 @@ public synchronized void setResourceOvercommit()
...
+
+        if (revocableFuture.isDone()) {
+            return future;
+        }
+        else {
+            return Futures.allAsList(ImmutableList.of(revocableFuture, future));
+        }
+    }
+
+    public synchronized boolean reserveRevocableMemory(long bytes)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: name it `tryReserve...`

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java [hunk 2]
```diff
@@ -50,6 +52,12 @@
     private long reserved;
 
     @GuardedBy("this")
+    private long revocableReserved;
+
+    @GuardedBy("this")
+    private SettableFuture<?> revocableFuture;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: Use simple `Future` here

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java [hunk 3]
```diff
@@ -62,6 +70,9 @@ public QueryContext(QueryId queryId, DataSize maxMemory, MemoryPool memoryPool,
         this.memoryPool = requireNonNull(memoryPool, "memoryPool is null");
         this.systemMemoryPool = requireNonNull(systemMemoryPool, "systemMemoryPool is null");
         this.executor = requireNonNull(executor, "executor is null");
+
+        this.revocableFuture = SettableFuture.create();
+        this.revocableFuture.set(null);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java:75]: use `NON_BLOCKED` here

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java [hunk 4]
```diff
@@ -158,4 +213,14 @@ public TaskContext addTaskContext(TaskStateMachine taskStateMachine, Session ses
         taskContexts.add(taskContext);
         return taskContext;
     }
+
+    private synchronized boolean exceedsMemoryLimit()
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: I would name it `isMemoryLimitExceeded`

File: presto-main/src/main/java/com/facebook/presto/operator/Driver.java [hunk 5]
```diff
@@ -324,10 +324,20 @@ private void processNewSource(TaskSource source)
                 // this state change by calling isFinished
                 return NOT_BLOCKED;
             }
-            return processInternal();
+            return processInternalAndHandleRevocableMemory();
         }
     }
 
+    private ListenableFuture<?> processInternalAndHandleRevocableMemory()
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: I don't like methods with `And` in the name. I would have another method `revokeMemoryIfNecessary` and the just call them both in a sequence

File: presto-main/src/main/java/com/facebook/presto/operator/DriverContext.java [hunk 6]
```diff
@@ -200,13 +201,27 @@ public void transferMemoryToTaskContext(long bytes)
...
+    }
+
     public ListenableFuture<?> reserveMemory(long bytes)
     {
         ListenableFuture<?> future = pipelineContext.reserveMemory(bytes);
         memoryReservation.getAndAdd(bytes);
         return future;
     }
 
+    public boolean reserveRevocableMemory(long bytes)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/DriverContext.java]: `tryReserve...`

File: presto-main/src/test/java/com/facebook/presto/memory/TestMemoryPools.java [hunk 7]
```diff
@@ -105,4 +175,76 @@ public static boolean isWaitingForMemory(List<Driver> drivers)
         }
         return false;
     }
+
+    private class RevocableMemoryOperator
+            implements Operator
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/memory/TestMemoryPools.java:181]: I would introduce an `AbstractOperator` which would have default implementation for most of these methods

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java (hunk 1 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: refactored

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java (hunk 2 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: I need `SettableFuture`

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java (hunk 3 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java:75]: dropped (after discussion)

File: presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java (hunk 4 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/DriverContext.java (hunk 6 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/DriverContext.java]: refactored

File: presto-main/src/test/java/com/facebook/presto/memory/TestMemoryPools.java (hunk 7 above)
Comment (pnowojski) [inline presto-main/src/test/java/com/facebook/presto/memory/TestMemoryPools.java:181]: dropped (after discussion)

File: presto-main/src/main/java/com/facebook/presto/operator/Driver.java (hunk 5 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/Driver.java]: done

**Comment by pnowojski** [comment]
Superseded by https://github.com/prestodb/presto/pull/5711

**Pull request closed without merging**